packages/flutter/lib/src/gestures/hit_test.dart

```diff
@@ -54,30 +61,188 @@ class HitTestEntry {
   /// be added in order from most specific to least specific, typically during an
   /// upward walk of the tree being hit tested.
   void add(HitTestEntry entry) {
-    _path.add(entry);
+    _path[entry] = _transforms.isEmpty ? null : _transforms.last;
+  }
+
+  /// Transforms `position` to the local coordinate system of a child before
+  /// hit-testing the child.
+  ///
+  /// The provided paint `transform` from the child coordinate system to the
+  /// coordinate system of the caller will be turned into a transform matrix
+  /// for pointer events. For this, the "perspective" component of the transform
+  /// is removed by calling [PointerEvent.paintTransformToPointerEventTransform].
+  /// The inverted transformed matrix is then used to transform `position` from
+  /// the parent coordinate system to the child coordinate system
+  /// before calling the provided `hitTest` callback, which needs to implement
+  /// the actual hit testing on the child. The [Offset] provided
+  /// to the `hitTest` callback is the position exactly under the user's
+  /// finger on the touch screen (or any other pointer device) in the local
+  /// coordinate system of the child.
+  ///
+  /// If `transform` is null it will be treated as the identity transform ad
+  /// `position` is provided to the `hitTest` callback as-is. If `transform`
+  /// cannot be inverted, the `hitTest` callback is not invoked and false is
+  /// returned. Otherwise, the return value of the `hitTest` callback is
+  /// returned.
+  ///
+  /// The `position` argument may be null, which will be forwarded to the
+  /// `hitTest` callback as-is. Using null as the position can be useful if
+  /// the child speaks a different hit test protocol then the parent and the
+  /// position is not required to do the actual hit testing in that protocol.
+  ///
+  /// {@tool sample}
+  /// This method is used in [RenderBox.hitTestChildren]:
+  ///
+  /// ```dart
+  /// abstract class Foo extends RenderBox {
+  ///
+  ///   final Matrix4 _effectiveTransform = Matrix4.rotationZ(50);
+  ///
+  ///   @override
+  ///   void applyPaintTransform(RenderBox child, Matrix4 transform) {
+  ///     transform.multiply(_effectiveTransform);
+  ///   }
+  ///
+  ///   @override
+  ///   bool hitTestChildren(HitTestResult result, { Offset position }) {
+  ///     return result.withTransform(
+  ///       transform: _effectiveTransform,
+  ///       position: position,
+  ///       hitTest: (HitTestResult result, Offset position) {
+  ///         return super.hitTestChildren(result, position: position);
+  ///       },
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///  * [withPaintOffset], which can be used for `transform`s that are just
+  ///    simple matrix translations by an [Offset].
+  ///  * [withRawTransform], which takes a transform matrix that is directly
+  ///    used to transform the position without any pre-processing.
+  bool withPaintTransform({
```

Comment:
i don't really understand what this does, from the name. It sounds like it returns a HitTestResult with a paint transform.

Reply:
Renamed these and the others to "addWithPaintTransform"